File: Medium/2616. Minimize the Maximum Difference of Pairs.java
class Solution {
    public int minimizeMax(int[] nums, int p) {
        final int N = nums.length;
        Arrays.sort(nums);
        int left = 0, right = nums[N - 1] - nums[0];
        while (left < right) {
            int mid = (left + right) / 2;
            int cnt = 0;
            for (int i = 1; i < N; i++) {
                if (nums[i] - nums[i - 1] <= mid) {
                    i++;
                    cnt++;
                }
            }
            if (cnt < p) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}

File: Medium/2616. Minimize the Maximum Difference of Pairs.py
from typing import List


class Solution:
    def minimizeMax(self, nums: List[int], p: int) -> int:
        if p == 0:
            return 0
        nums.sort()
        n, lo, hi = len(nums), 0, nums[-1] - nums[0]
        while lo < hi:
            mi, pairs = (lo + hi) // 2, 0
            i = 1
            while i < n:
                if nums[i] - nums[i - 1] <= mi:
                    pairs += 1
                    i += 1
                i += 1
            if pairs >= p:
                hi = mi
            else:
                lo = mi + 1
        return lo